File: android/src/main/kotlin/com/cloudwebrtc/webrtc/model/IceConnectionState.kt
package com.cloudwebrtc.webrtc.model

import org.webrtc.PeerConnection.IceConnectionState as WIceConnectionState

/**
 * Representation of the [org.webrtc.PeerConnection.IceConnectionState].
 *
 * @property value [Int] representation of this enum which will be expected on Flutter side.
 */
enum class IceConnectionState(val value: Int) {
    NEW(0),
    CHECKING(1),
    CONNECTED(2),
    COMPLETED(3),
    FAILED(4),
    DISCONNECTED(5),
    CLOSED(6);

    companion object {
        /**
         * Converts provided [org.webrtc.PeerConnection.IceConnectionState] into [IceConnectionState].
         *
         * @return [IceConnectionState] created based on the provided [org.webrtc.PeerConnection.IceConnectionState].
         */
        fun fromWebRtc(from: WIceConnectionState): IceConnectionState {
            return when (from) {
                WIceConnectionState.NEW -> NEW
                WIceConnectionState.CHECKING -> CHECKING
                WIceConnectionState.CONNECTED -> CONNECTED
                WIceConnectionState.COMPLETED -> COMPLETED
                WIceConnectionState.FAILED -> FAILED
                WIceConnectionState.DISCONNECTED -> DISCONNECTED
                WIceConnectionState.CLOSED -> CLOSED
            }
        }
    }
}

File: ios/Classes/model/IceConnectionState.swift
import WebRTC

public enum IceConnectionState: Int {
  case new, checking, connected, completed, failed, disconnected, closed

  static func fromWebRtc(state: RTCIceConnectionState) -> IceConnectionState {
    switch state {
    case .new:
      return IceConnectionState.new
    case .checking:
      return IceConnectionState.checking
    case .connected:
      return IceConnectionState.connected
    case .completed:
      return IceConnectionState.completed
    case .failed:
      return IceConnectionState.failed
    case .disconnected:
      return IceConnectionState.disconnected
    case .closed:
      return IceConnectionState.closed
    case .count:
      // TODO: in Rust impl it's marked just unreachable. Why???
      abort()
    }
  }

  func asFlutterResult() -> Int {
    return self.rawValue
  }
}
